mirgecom/wave.py

```diff
@@ -23,70 +23,71 @@
 import numpy as np
 import numpy.linalg as la  # noqa
 from pytools.obj_array import (
-    join_fields, make_obj_array,
-    with_object_array_or_scalar)
+    flat_obj_array, make_obj_array)
 from meshmode.mesh import BTAG_ALL, BTAG_NONE  # noqa
-from grudge.eager import with_queue
+from meshmode.dof_array import thaw
 from grudge.symbolic.primitives import TracePair
+from grudge.eager import interior_trace_pair
 
 
 __doc__ = """
 .. autofunction:: wave_operator
 """
 
 
-def _interior_trace_pair(discr, vec):
-    i = discr.interp("vol", "int_faces", vec)
-    e = with_object_array_or_scalar(
-        lambda el: discr.opposite_face_connection()(el.queue, el),
-        i)
-    return TracePair("int_faces", i, e)
-
-
-def _flux(discr, c, w_tpair):
+def _flux(actx, discr, c, w_tpair):
     u = w_tpair[0]
     v = w_tpair[1:]
 
-    normal = with_queue(u.int.queue, discr.normal(w_tpair.dd))
+    normal = thaw(actx, discr.normal(w_tpair.dd))
 
     def normal_times(scalar):
         # workaround for object array behavior
         return make_obj_array([ni*scalar for ni in normal])
 
-    flux_weak = join_fields(
+    flux_weak = flat_obj_array(
         np.dot(v.avg, normal),
         normal_times(u.avg),
         )
 
     # upwind
     v_jump = np.dot(normal, v.int-v.ext)
-    flux_weak += join_fields(
+    flux_weak += flat_obj_array(
         0.5*(u.int-u.ext),
         0.5*normal_times(v_jump),
         )
 
-    return discr.interp(w_tpair.dd, "all_faces", c*flux_weak)
+    return discr.project(w_tpair.dd, "all_faces", c*flux_weak)
+
 
+def wave_operator(actx, discr, c, w):
+    """
+    Args:
```

Comment:
Probably want to add argument documentation for `actx` here?

Would it make any sense to store a reference to the array context inside the discretization object, in order to avoid always having to pass both? (I don't have a good sense for exactly how these two relate yet...)

Reply:
Good catch! Passing an (extra) array context really doesn't make any sense, as the state vector/trace pair already have one on board. Fixed in dde43d4.